src/os/bluestore/BlueFS.cc

```diff
@@ -1889,7 +1918,17 @@ void BlueFS::sync_metadata()
   flush_bdev(); // FIXME?
   _flush_and_sync_log(l);
   for (unsigned i = 0; i < to_release.size(); ++i) {
+    int r = 0;
+    if (cct->_conf->bdev_enable_discard && cct->_conf->bdev_async_discard) {
+      if (!to_release[i].empty())
+        r = bdev[i]->queue_discard(to_release[i]);
+      if (r == 0)
+        continue;
+    }
+
     for (auto p = to_release[i].begin(); p != to_release[i].end(); ++p) {
+      if (cct->_conf->bdev_enable_discard && r == 0)
+        bdev[i]->discard(p.get_start(), p.get_len());
```

Comment:
I think this code is also wrong, because r is undefined in that scope...
The intention of the code is that it discards given range which is failed to discard async, synchronously.

And i don't understand your(@liewegas) comment "This doesn't seem right.. first, if you discard async, you do'nt want ot later discard() sync, or call release(), right?"
Does it mean 'Don't discard synchronously and don't call release(), if discard has been done async'
Am i right??

Reply:
That's a comment I deleted after I looked harder and understood the code :)